### PR TITLE
Fix flash of unstyled content

### PR DESCRIPTION
**Pull request by benhalpern**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

There is an bug with the timing of how all the CSS is loaded from cold. There is evidence that Google is detecting this as a cumulative layout shift problem that needs to be adjusted, and anecdotally I have run into this occasionally myself.

More details in the linked issue, but here's the visual:

https://user-images.githubusercontent.com/3102842/105424035-8ecf8c80-5c14-11eb-9eaa-d5f8653ce43b.mov

I speculate that all the variables we use in CSS trips up the content rendering tree and we get a flash of the "user agent styling" mid-way through rendering the content. It only shows up for non-logged-in traffic, and I believe this is due to the fact that we run less blocking JS on non-logged-in traffic (a good thing). Signed in traffic incidentally gets other render-blocking resources to hide the content tree mishaps.

My solution is to include the stylesheet links in both the head and the body. The browser will know not to fetch them both from source, so it should not cause extra delays, but this seems to alter the paint timing such that users and SEO bots do not get tripped up with this glitch.

I modified the "asset reconciliation" functionality to ensure if we append additional stylesheets they will land after both stylesheets. I also modified the logic to only run this for _signed in traffic_. Ultimately this is not related to the fix, but it reduces the amount of moving pieces. "Asset reconciliation" is designed to help fix mismatches between cached shell and body content which is much more likely to occur for repeat signed-in traffic. We could revisit this another time.

This is definitely just a patch which can be supplemented by further investigation into refactoring of our CSS to get a better handle on things, but this hack doesn't add any great additional code complexity.

## Related Tickets & Documents

Closes https://github.com/forem/forem/issues/12377

## QA Instructions, Screenshots, Recordings

Details in the issue.

### UI accessibility concerns?

None that I am aware of.

## Added tests?

- [x] Yes - modified test.
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

### CODE REVIEW

File: app/views/shell/_top.html.erb
Comment (benhalpern) [inline app/views/shell/_top.html.erb:21]: This is an old hack we can remove. ✂️

(If I'm adding a bit of a hack to fix this, best to balance the universe and remove one too)

File: app/views/layouts/_styles.html.erb
Comment (benhalpern) [inline app/views/layouts/_styles.html.erb:1]: These comments have served their purpose and can be laid to rest.

File: app/views/shell/_top.html.erb
Comment (benhalpern) [inline app/views/shell/_top.html.erb:59]: This is the main functionality of this patch. This partial is now in the `head` _and_ the top of `body`.

**Review by atsmith813 — APPROVED** [review]

**Comment by mstruve** [comment]
Given this PR is doing multiple things and the bug does not appear urgent to me I would like to hold off merging this. I think we should take a beat and make an issue and then maybe one of our awesome new frontend devs can help make sure we are approaching this correctly and putting in a solid long term fix.

**Comment by benhalpern** [comment]
@mstruve I'm fine to let this sit since it's not _ultra urgent_.

I do think this isn't so much "doing multiple things"... Only one line change is of real consequence.

File: app/views/shell/_top.html.erb
Comment (ludwiczakpawel) [reply app/views/shell/_top.html.erb:21]: Can't wait do drop that `<style>` from there - its been driving me crazy ever since

**Review by ludwiczakpawel — APPROVED** [review]
I approve after reviewing the parts of the code I can understand as well as after testing things locally 👍

**Review by katiedavis — APPROVED** [review]

**Review by mstruve — APPROVED** [review]
@benhalpern given we got a lot of good eyes on this from the team I think we are good to merge. Thank you for taking a pause to make sure this was a solution that not only worked but would be sustainable going forward. I also really appreciate you making the issue for it.

**Pull request merged**
